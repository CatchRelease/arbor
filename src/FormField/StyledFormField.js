import styled from '@emotion/styled';
import PropTypes from 'prop-types';

const StyledFormField = styled.div`
  margin-bottom: ${props =>
    props.caption ? props.theme.space.smaller : props.theme.space.regular};
`;

StyledFormField.propTypes = {
  caption: PropTypes.string
};

StyledFormField.defaultProps = {
  caption: ''
};

export default StyledFormField;
