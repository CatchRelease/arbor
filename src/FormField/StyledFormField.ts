import styled from '@emotion/styled';

type Props = {
  caption?: string | null;
};

const StyledFormField = styled.div<Props>`
  margin-bottom: ${(props) =>
    props.caption ? props.theme.space.smaller : props.theme.space.regular};
`;

StyledFormField.defaultProps = {
  caption: ''
};

export default StyledFormField;
