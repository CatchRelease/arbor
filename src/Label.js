import styled from 'react-emotion';
import PropTypes from 'prop-types';

const Label = styled.label`
  color: ${props => props.theme.colors.grey80};
  display: block;
  font-family: ${props => props.theme.brandFont};
  font-size: ${props => props.theme.fontSizes.size2};
  line-height: ${props => props.theme.lineHeights.large};
`;

Label.propTypes = {
  htmlFor: PropTypes.string.isRequired
};

export default Label;
