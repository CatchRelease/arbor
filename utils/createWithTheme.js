import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
