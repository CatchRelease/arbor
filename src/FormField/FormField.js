import React from 'react';
import PropTypes from 'prop-types';

import Label from '../Label';
import Paragraph from '../Paragraph';
import StyledFormField from './StyledFormField';

const FormField = ({ caption, children, label, id }) => (
  <StyledFormField caption={caption}>
    {label && <Label htmlFor={id}>{label}</Label>}
    {children}
    {caption && (
      <Paragraph.span variant="tiny" color="red">
        {caption}
      </Paragraph.span>
    )}
  </StyledFormField>
);

FormField.propTypes = {
  caption: PropTypes.string,
  children: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string
};

FormField.defaultProps = {
  caption: '',
  label: ''
};

export default FormField;
