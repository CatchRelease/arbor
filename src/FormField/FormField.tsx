import { FC, ReactNode } from 'react';

import Flex from '../Flex';
import Label from '../Label';
import Paragraph from '../Paragraph';
import StyledFormField from './StyledFormField';

type Props = {
  caption?: string | null;
  id: string;
  label?: ReactNode | string;
  labelAside?: ReactNode;
};

const FormField: FC<Props> = ({
  caption = '',
  children,
  label = '',
  labelAside,
  id
}) => (
  <StyledFormField caption={caption}>
    <Flex alignItems="center">
      {label && (
        <Flex flex="1" alignItems="center">
          <Label htmlFor={id}>{label}</Label>
        </Flex>
      )}
      {labelAside && (
        <Flex color="text.muted" fontSize="size4">
          {labelAside}
        </Flex>
      )}
    </Flex>
    {children}
    {caption && (
      <Paragraph as="span" variant="tiny" color="palette.red.default">
        {caption}
      </Paragraph>
    )}
  </StyledFormField>
);

export default FormField;
