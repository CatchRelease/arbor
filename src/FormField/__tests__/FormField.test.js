import { shallow } from 'enzyme';

import createWithTheme from '../../../utils/createWithTheme';
import Label from '../../Label';
import FormField from '../FormField';

describe('<FormField />', () => {
  it('renders a FormField correctly', () => {
    const tree = createWithTheme(
      <FormField id="exmaple">Field Child</FormField>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders a FormField with a caption correctly', () => {
    const tree = createWithTheme(
      <FormField caption="I am a caption" id="example">
        Field Child
      </FormField>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  context('with label', () => {
    it('renders a FormField with a label correctly', () => {
      const tree = createWithTheme(
        <FormField id="example" label="My label">
          Field Child
        </FormField>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('associates the id with the label', () => {
      const formField = shallow(
        <FormField id="example" label="My label">
          Field Child
        </FormField>
      );

      expect(formField.find(Label).props().htmlFor).toEqual('example');
    });
  });

  context('with labelAside', () => {
    it('renders a FormField with a label correctly', () => {
      const labelAside = <div id="my-label-aside">muffins</div>;

      const tree = createWithTheme(
        <FormField id="example" label="My label" labelAside={labelAside}>
          Field Child
        </FormField>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
