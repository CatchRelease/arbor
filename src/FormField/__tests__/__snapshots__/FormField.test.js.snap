// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> renders a FormField correctly 1`] = `
.emotion-0 {
  margin-bottom: 1rem;
}

<div
  className="emotion-0 emotion-1"
>
  
  Field Child
  
</div>
`;

exports[`<FormField /> renders a FormField with a caption correctly 1`] = `
.emotion-2 {
  margin-bottom: 0.5rem;
}

.emotion-0 {
  font-family: Motiva Sans;
  color: #EE2200;
  line-height: 1.25;
  font-size: 0.75rem;
}

<div
  className="emotion-2 emotion-3"
>
  
  Field Child
  <span
    className="emotion-0 emotion-1"
    color="red"
  >
    I am a caption
  </span>
</div>
`;

exports[`<FormField /> with label renders a FormField with a label correctly 1`] = `
.emotion-2 {
  margin-bottom: 1rem;
}

.emotion-0 {
  color: #5B5D61;
  display: block;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.5;
}

<div
  className="emotion-2 emotion-3"
>
  <label
    className="emotion-0 emotion-1"
    htmlFor="example"
  >
    My label
  </label>
  Field Child
  
</div>
`;
